Framework/Algorithms/src/MergeRuns/SampleLogsBehaviour.cpp

```diff
@@ -0,0 +1,463 @@
+ *merge
+ * @param addeeWSNumber a double for the addeeWS value (0 if it is not numeric)
+ * @param outWSNumber a double for the outWS value (0 if it is not numeric)
+ * @param name the name of the sample log to check
+ */
+void SampleLogsBehaviour::checkWarnProperty(const MatrixWorkspace_sptr &addeeWS,
+                                            Property *addeeWSProperty,
+                                            const SampleLogBehaviour &behaviour,
+                                            const double addeeWSNumber,
+                                            const double outWSNumber,
+                                            const std::string name) {
+  if (behaviour.isNumeric && behaviour.tolerance > 0.0) {
+    if (std::abs(addeeWSNumber - outWSNumber) > behaviour.tolerance) {
+      m_logger.warning() << generateDifferenceMessage(
+          name, addeeWS->name(), addeeWSProperty->value(),
+          behaviour.property->value());
+    }
+  } else {
+    if (behaviour.property->value().compare(addeeWSProperty->value()) != 0) {
+      m_logger.warning() << generateDifferenceMessage(
+          name, addeeWS->name(), addeeWSProperty->value(),
+          behaviour.property->value());
+    }
+  }
+}
+
+/**
+ ** Performs the check to see if an error should be generated because logs are
+ *different. Performs a numeric comparison if a tolerance is set and the log is
+ *a number, else performs a string comparison.
+ *
+ * @param addeeWS the workspace being merged
+ * @param addeeWSProperty the property value of the workspace being merged
+ * @param behaviour the merge behaviour struct, containing information about the
+ *merge
+ * @param addeeWSNumber a double for the addeeWS value (0 if it is not numeric)
+ * @param outWSNumber a double for the outWS value (0 if it is not numeric)
+ * @param name the name of the sample log to check
+ */
+void SampleLogsBehaviour::checkErrorProperty(
+    const MatrixWorkspace_sptr &addeeWS, Property *addeeWSProperty,
+    const SampleLogBehaviour &behaviour, const double addeeWSNumber,
+    const double outWSNumber, const std::string name) {
+  if (behaviour.isNumeric && behaviour.tolerance > 0.0) {
+    if (std::abs(addeeWSNumber - outWSNumber) > behaviour.tolerance) {
+      throw std::invalid_argument(generateDifferenceMessage(
+          name, addeeWS->name(), addeeWSProperty->value(),
+          behaviour.property->value()));
+    }
+  } else {
+    if (behaviour.property->value().compare(addeeWSProperty->value()) != 0) {
+      throw std::invalid_argument(generateDifferenceMessage(
+          name, addeeWS->name(), addeeWSProperty->value(),
+          behaviour.property->value()));
+    }
+  }
+}
+
+/**
+ * Set the values in the map to be the same as those in the output workspace.
+ *
+ * @param ws the merged workspace
+ */
+void SampleLogsBehaviour::setUpdatedSampleLogs(
+    const API::MatrixWorkspace_sptr &ws) {
+  for (auto item : m_logMap) {
```

Comment:
Should this be a reference? `item` is modified below, but the value is a `shared_ptr`, so it is ok... but it is confusing and inefficient.